tls/src/keys.rs

```diff
@@ -0,0 +1,362 @@
+    }
+
+    pub fn load_certified_key(&self) -> Result<Arc<CertifiedKey>, Error> {
+        match &self.resolve {
+            ResolveSetting::Ipcc => self.load_ipcc_key(),
+            ResolveSetting::Local {
+                priv_key,
+                cert_chain,
+            } => self.load_local_key(priv_key, cert_chain),
+        }
+    }
+}
+
+/// A mechanism for signing using an in memory Ed25519 private key
+#[derive(Debug)]
+pub(crate) struct LocalEd25519Signer {
+    // Not necessary to wrap in a `Secret`. It already prevents debug printing
+    // of the secret.
+    key: ed25519_dalek::SigningKey,
+}
+
+impl Signer for LocalEd25519Signer {
+    fn sign(&self, message: &[u8]) -> Result<Vec<u8>, rustls::Error> {
+        // XXX double check on the use of sha3-256 because that's what we need to use with
+        // the RoT
+        let mut prehashed = sha3::Sha3_256::new();
+        prehashed.update(message);
+        let sig = self.key.sign(&prehashed.finalize());
+
+        Ok(sig.to_vec())
+    }
+
+    fn scheme(&self) -> SignatureScheme {
+        SignatureScheme::ED25519
+    }
+}
+
+/// A DER encoded private key
+#[derive(Debug, Zeroize, ZeroizeOnDrop)]
+pub struct PrivkeyDer(Vec<u8>);
+
+impl DebugSecret for PrivkeyDer {}
+
+/// An implementation of a an Ed25519 private signing key that lives in memory
+///
+/// In production we'll send signing requests to the RoT via IPCC and sprot.
+#[derive(Debug)]
+pub(crate) struct LocalEd25519SigningKey {
+    privkey_der: Secret<PrivkeyDer>,
+}
+
+impl SigningKey for LocalEd25519SigningKey {
+    fn choose_scheme(
+        &self,
+        offered: &[SignatureScheme],
+    ) -> Option<Box<dyn rustls::sign::Signer>> {
+        if !offered.iter().any(|&s| s == SignatureScheme::ED25519) {
+            return None;
+        }
+
+        let privkey_info = PrivateKeyInfo::try_from(
+            self.privkey_der.expose_secret().0.as_slice(),
+        )
+        .ok()?;
+
+        let signing_key =
+            ed25519_dalek::SigningKey::try_from(privkey_info).ok()?;
+
+        Some(Box::new(LocalEd25519Signer { key: signing_key })
+            as Box<dyn Signer>)
+    }
+
+    fn algorithm(&self) -> rustls::SignatureAlgorithm {
+        rustls::SignatureAlgorithm::ED25519
+    }
+}
+
+/// Represents the underlying key returned over IPCC
+#[derive(Debug)]
+pub struct IpccKey {}
+
+impl SigningKey for IpccKey {
+    fn choose_scheme(
+        &self,
+        offered: &[SignatureScheme],
+    ) -> Option<Box<dyn rustls::sign::Signer>> {
+        if !offered.iter().any(|&s| s == SignatureScheme::ED25519) {
+            return None;
+        }
+        Some(Box::new(IpccSigner {}) as Box<dyn Signer>)
+    }
+
+    fn algorithm(&self) -> rustls::SignatureAlgorithm {
+        rustls::SignatureAlgorithm::ED25519
+    }
+}
+
+#[derive(Debug)]
+pub struct IpccSigner {}
+
+impl Signer for IpccSigner {
+    fn sign(&self, message: &[u8]) -> Result<Vec<u8>, rustls::Error> {
+        // We require sha3_256
```

Comment:
If we are hashing the message before sending it down and verifying the hash, and both are done in code here on the host, we can likely chose whatever algorithm we want right? I think the issue with Sha512 is that maybe we expect a 256 bit sized hash on the RoT, but the RoT has no idea whether that hash was generated by sha256 or Sha3-256, right?

In any case, I think using sha3_256 is the right move.